quicktill/pdrivers.py

```diff
@@ -689,6 +693,9 @@ def process_canvas(self, canvas, f):
                         ("%s%s%s%s%s\n" % (
                             left, ' ' * padl, center, ' ' * padr, right))
                         .encode(self.coding))
+            elif hasattr(i, 'image_data'):
+                assert len(i.image_data * 8) == i.image_width * i.image_height
```

Comment:
This assertion is failing when I test this. I generated my PBM file using `convert icon-512x512.png -colors 2 -colorspace gray emflogo.pbm` — is this correct?

Reply:
My PBM file starts:
```
P4
512 512
```
And it's 32779 bytes long.

Reply:
Adding a log.debug() just before the assert shows:
```
len(i.image_data)=32770 i.image_width=51 i.image_height=5
```
(You probably want to be doing `len(i.image_data) * 8` rather than `len(i.image_data * 8)`!)

Reply:
Thanks - yep, that's some fairly awful code I've written there.  Assuming two-digit (character) width and height in the file header.  I'll go check whether there's some existing functionality I can re-use to parse the file header.

Reply:
Also: I have a vague memory that I had more success with the `-monochrome` flag to `convert` than `-colors 2`; but it did kinda seem that they should behave the same.

Reply:
I've attempted to address these, but haven't yet re-tested the results (so it's quite likely I've missed something).  I'll test the results within the next day or so.

Reply:
Ok; from re-testing: a receipt output is printed (no further exceptions occurred), but (as expected) it still has layout/formatting problems.

Reply:
I think the vertical formatting problem (empty space between pixels in the image) during my testing may be something to do with the fact that [`ESC *` mode zero](https://download4.epson.biz/sec_pubs/pos/reference_en/escpos/esc_asterisk.html) expects 8 bits of vertical data.